File: internal/terminal/profile_config_test.go
package terminal

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

func TestValidateProfileNormalizesSafePresentationAndPolicyDefaults(t *testing.T) {
	input := configuredTestProfile(t, "shell-default", ProfileShell)
	input.Args = []string{"-l"}
	input.Env = map[string]string{"PAGER": "less"}

	got, err := ValidateProfile(input)
	if err != nil {
		t.Fatalf("ValidateProfile: %v", err)
	}
	if got.Theme != DefaultProfileTheme || got.FontFamily != DefaultProfileFontFamily ||
		got.FontSize != DefaultProfileFontSize || got.Scrollback != DefaultProfileScrollback ||
		got.CWDPolicy != CWDRequested || got.FixedCWD != "" || got.ExitBehavior != ExitKeep {
		t.Fatalf("normalized profile defaults = %#v", got)
	}
	if input.Theme != "" || input.FontFamily != "" || input.FontSize != 0 ||
		input.Scrollback != 0 || input.CWDPolicy != "" || input.ExitBehavior != "" {
		t.Fatalf("ValidateProfile mutated caller defaults: %#v", input)
	}

	input.Args[0] = "changed"
	input.Env["PAGER"] = "changed"
	if got.Args[0] != "-l" || got.Env["PAGER"] != "less" {
		t.Fatalf("validated profile aliases caller data: %#v", got)
	}
}

func TestValidateProfileBoundsSettingsAndRejectsAuthorityOrSecretEnvironment(t *testing.T) {
	fixed := t.TempDir()
	valid := configuredTestProfile(t, "profile", ProfileShell)
	valid.Theme = "solarized-dark"
	valid.FontFamily = "Iosevka, monospace"
	valid.FontSize = MinProfileFontSize
	valid.Scrollback = MaxProfileScrollback
	valid.CWDPolicy = CWDFixed
	valid.FixedCWD = fixed
	valid.ExitBehavior = ExitCloseOnSuccess
	if _, err := ValidateProfile(valid); err != nil {
		t.Fatalf("ValidateProfile valid settings: %v", err)
	}

	tests := []struct {
		name   string
		change func(*Profile)
	}{
		{name: "theme", change: func(profile *Profile) { profile.Theme = "not a stable theme" }},
		{name: "font size", change: func(profile *Profile) { profile.FontSize = MaxProfileFontSize + 1 }},
		{name: "scrollback", change: func(profile *Profile) { profile.Scrollback = MaxProfileScrollback + 1 }},
		{name: "relative fixed CWD", change: func(profile *Profile) {
			profile.CWDPolicy, profile.FixedCWD = CWDFixed, "relative"
		}},
		{name: "unused fixed CWD", change: func(profile *Profile) {
			profile.CWDPolicy, profile.FixedCWD = CWDProject, fixed
		}},
		{name: "exit behavior", change: func(profile *Profile) { profile.ExitBehavior = "restart" }},
		{name: "argument count", change: func(profile *Profile) {
			profile.Args = make([]string, maxProfileArgumentCount+1)
		}},
		{name: "reserved environment", change: func(profile *Profile) {
			profile.Env = map[string]string{"PTRACK_CAPABILITY_TOKEN": "value"}
		}},
		{name: "case-insensitive reserved environment", change: func(profile *Profile) {
			profile.Env = map[string]string{"ptrack_custom": "value"}
		}},
		{name: "secret environment", change: func(profile *Profile) {
			profile.Env = map[string]string{"OPENAI_API_KEY": "value"}
		}},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			profile := configuredTestProfile(t, "profile", ProfileShell)
			test.change(&profile)
			if _, err := ValidateProfile(profile); err == nil {
				t.Fatal("ValidateProfile accepted invalid bounded configuration")
			}
		})
	}
}

func TestMergeProfilesReplacesStableIDsAddsCustomProfilesAndOwnsCopies(t *testing.T) {
	discoveredShell := configuredTestProfile(t, "shell-default", ProfileShell)
	discoveredShell.Name = "Default shell"
	discoveredShell.Args = []string{"-l"}
	discoveredShell.Env = map[string]string{"PAGER": "less"}
	discoveredAgent := configuredTestProfile(t, "agent-codex", ProfileAgent)
	discoveredAgent.Name = "Codex"
	discoveredAgent.Provider = "codex"
	discoveredAgent.Args = []string{"--discovered"}
	discoveredAgent.Env = map[string]string{"CODEX_MODE": "base"}

	override := discoveredAgent
	override.Name = "Codex focused"
	override.Theme = "solarized-dark"
	override.FontSize = MaxProfileFontSize
	override.ExitBehavior = ExitCloseOnSuccess
	custom := configuredTestProfile(t, "shell-tools", ProfileShell)
	custom.Name = "Tools shell"
	custom.Args = []string{"--custom"}
	custom.Env = map[string]string{"PAGER": "more"}

	merged, err := MergeProfiles(
		[]Profile{discoveredAgent, discoveredShell},
		[]Profile{custom, override},
	)
	if err != nil {
		t.Fatalf("MergeProfiles: %v", err)
	}
	wantIDs := []string{"shell-default", "shell-tools", "agent-codex"}
	gotIDs := make([]string, 0, len(merged))
	for _, profile := range merged {
		gotIDs = append(gotIDs, profile.ID)
	}
	if !reflect.DeepEqual(gotIDs, wantIDs) {
		t.Fatalf("merged IDs = %v, want %v", gotIDs, wantIDs)
	}
	gotAgent := merged[2]
	if gotAgent.Name != override.Name || !reflect.DeepEqual(gotAgent.Args, discoveredAgent.Args) ||
		gotAgent.Env["CODEX_MODE"] != "base" || gotAgent.Theme != "solarized-dark" ||
		gotAgent.FontSize != MaxProfileFontSize || gotAgent.ExitBehavior != ExitCloseOnSuccess {
		t.Fatalf("configured override = %#v", gotAgent)
	}
	if !reflect.DeepEqual(merged[1].Args, custom.Args) || merged[1].Env["PAGER"] != "more" {
		t.Fatalf("custom shell launch settings = %#v", merged[1])
	}

	override.Args[0] = "changed"
	override.Env["CODEX_MODE"] = "changed"
	custom.Args[0] = "changed"
	custom.Env["PAGER"] = "changed"
	discoveredShell.Args[0] = "changed"
	discoveredShell.Env["PAGER"] = "changed"
	if gotAgent.Args[0] != "--discovered" || gotAgent.Env["CODEX_MODE"] != "base" ||
		merged[0].Args[0] != "-l" || merged[0].Env["PAGER"] != "less" ||
		merged[1].Args[0] != "--custom" || merged[1].Env["PAGER"] != "more" {
		t.Fatalf("merged profiles alias input data: %#v", merged)
	}
}

func TestMergeProfilesRejectsDuplicateAndRepurposedIdentities(t *testing.T) {
	discovered := configuredTestProfile(t, "agent-codex", ProfileAgent)
	discovered.Provider = "codex"

	duplicate := configuredTestProfile(t, "custom", ProfileShell)
	if _, err := MergeProfiles(nil, []Profile{duplicate, duplicate}); err == nil {
		t.Fatal("MergeProfiles accepted duplicate configured IDs")
	}

	changedKind := discovered
	changedKind.Kind = ProfileShell
	changedKind.Provider = ""
	if _, err := MergeProfiles([]Profile{discovered}, []Profile{changedKind}); err == nil {
		t.Fatal("MergeProfiles accepted a changed discovered kind")
	}

	changedProvider := discovered
	changedProvider.Provider = "other"
	if _, err := MergeProfiles([]Profile{discovered}, []Profile{changedProvider}); err == nil {
		t.Fatal("MergeProfiles accepted a changed discovered provider")
	}

	changedExecutable := discovered
	changedExecutable.Executable = filepath.Join(t.TempDir(), "other")
	if _, err := MergeProfiles([]Profile{discovered}, []Profile{changedExecutable}); err == nil {
		t.Fatal("MergeProfiles accepted a changed discovered agent executable")
	}

	changedArgs := discovered
	changedArgs.Args = []string{"--changed"}
	if _, err := MergeProfiles([]Profile{discovered}, []Profile{changedArgs}); err == nil {
		t.Fatal("MergeProfiles accepted changed discovered agent arguments")
	}

	changedEnvironment := discovered
	changedEnvironment.Env = map[string]string{"DYLD_INSERT_LIBRARIES": "/tmp/injected.dylib"}
	if _, err := MergeProfiles([]Profile{discovered}, []Profile{changedEnvironment}); err == nil {
		t.Fatal("MergeProfiles accepted changed discovered agent environment")
	}

	changedCWD := discovered
	changedCWD.CWDPolicy = CWDProject
	if _, err := MergeProfiles([]Profile{discovered}, []Profile{changedCWD}); err == nil {
		t.Fatal("MergeProfiles accepted changed discovered agent working-directory policy")
	}

	fixedCWD := discovered
	fixedCWD.CWDPolicy = CWDFixed
	fixedCWD.FixedCWD = t.TempDir()
	changedFixedCWD := fixedCWD
	changedFixedCWD.FixedCWD = t.TempDir()
	if _, err := MergeProfiles([]Profile{fixedCWD}, []Profile{changedFixedCWD}); err == nil {
		t.Fatal("MergeProfiles accepted changed discovered agent fixed working directory")
	}

	customAgent := configuredTestProfile(t, "agent-custom", ProfileAgent)
	if _, err := MergeProfiles(nil, []Profile{customAgent}); err == nil {
		t.Fatal("MergeProfiles accepted a custom configured agent identity")
	}
}

func TestProfileConfigPrivateAtomicRoundTripDoesNotSnapshotInheritedEnvironment(t *testing.T) {
	directory := t.TempDir()
	path := filepath.Join(directory, "profiles.json")
	t.Setenv("PROFILE_CONFIG_PARENT_SECRET", "must-not-persist")

	first := configuredTestProfile(t, "shell-default", ProfileShell)
	first.Name = "First"
	first.Env = map[string]string{"PAGER": "less"}
	config := ProfileConfig{Version: ProfileConfigVersion, Profiles: []Profile{first}}
	if err := SaveProfileConfig(path, config); err != nil {
		t.Fatalf("SaveProfileConfig first: %v", err)
	}

	second := first
	second.Name = "Second"
	second.Env = map[string]string{"PAGER": "more"}
	if err := SaveProfileConfig(path, ProfileConfig{
		Version: ProfileConfigVersion, Profiles: []Profile{second},
	}); err != nil {
		t.Fatalf("SaveProfileConfig replacement: %v", err)
	}

	assertProfileConfigPrivate(t, path)
	contents, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read profile config: %v", err)
	}
	if strings.Contains(string(contents), "PROFILE_CONFIG_PARENT_SECRET") ||
		strings.Contains(string(contents), "must-not-persist") {
		t.Fatal("profile config snapshots inherited environment")
	}
	entries, err := os.ReadDir(directory)
	if err != nil {
		t.Fatalf("read config directory: %v", err)
	}
	if len(entries) != 1 || entries[0].Name() != filepath.Base(path) {
		t.Fatalf("atomic save left temporary files: %v", entries)
	}

	loaded, err := LoadProfileConfig(path)
	if err != nil {
		t.Fatalf("LoadProfileConfig: %v", err)
	}
	if loaded.Version != ProfileConfigVersion || len(loaded.Profiles) != 1 ||
		loaded.Profiles[0].Name != "Second" || loaded.Profiles[0].Env["PAGER"] != "more" ||
		loaded.Profiles[0].Theme != DefaultProfileTheme {
		t.Fatalf("loaded profile config = %#v", loaded)
	}

	second.Env["PAGER"] = "changed"
	if loaded.Profiles[0].Env["PAGER"] != "more" {
		t.Fatal("loaded profile config aliases save input")
	}
}

func TestLoadProfileConfigRejectsUnknownVersionFieldsTrailingAndOversize(t *testing.T) {
	directory := t.TempDir()
	profile := configuredTestProfile(t, "shell", ProfileShell)
	profileJSON, err := json.Marshal(profile)
	if err != nil {
		t.Fatal(err)
	}
	tests := []struct {
		name string
		body string
	}{
		{name: "version", body: `{"version":2,"profiles":[]}`},
		{name: "unknown", body: `{"version":1,"profiles":[],"extra":true}`},
		{name: "trailing", body: `{"version":1,"profiles":[]} {}`},
		{name: "invalid profile", body: `{"version":1,"profiles":[` + string(profileJSON[:len(profileJSON)-1]) + `]}`},
		{name: "oversize", body: strings.Repeat(" ", maxProfileConfigJSONBytes+1)},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			path := filepath.Join(directory, test.name+".json")
			if err := os.WriteFile(path, []byte(test.body), 0o600); err != nil {
				t.Fatal(err)
			}
			if _, err := LoadProfileConfig(path); err == nil {
				t.Fatal("LoadProfileConfig accepted invalid input")
			}
		})
	}
}

func TestManagerEnforcesProfileWorkingDirectoryPolicy(t *testing.T) {
	projectRoot := t.TempDir()
	requested := t.TempDir()
	fixed := t.TempDir()
	projectProcess := newManagerFakeProcess()
	fixedProcess := newManagerFakeProcess()
	factory := newManagerFakeFactory(
		managerStartOutcome{process: projectProcess},
		managerStartOutcome{process: fixedProcess},
	)
	projectProfile := configuredTestProfile(t, "shell-project", ProfileShell)
	projectProfile.CWDPolicy = CWDProject
	fixedProfile := configuredTestProfile(t, "shell-fixed", ProfileShell)
	fixedProfile.CWDPolicy = CWDFixed
	fixedProfile.FixedCWD = fixed
	manager, err := NewManager(projectRoot, []Profile{projectProfile, fixedProfile}, factory)
	if err != nil {
		t.Fatalf("NewManager: %v", err)
	}
	cleanupManager(t, manager, projectProcess, fixedProcess)

	if _, err := manager.Create(projectProfile.ID, requested, 24, 80); err != nil {
		t.Fatalf("create project-policy terminal: %v", err)
	}
	if _, err := manager.Create(fixedProfile.ID, requested, 24, 80); err != nil {
		t.Fatalf("create fixed-policy terminal: %v", err)
	}
	starts := factory.recordedStarts()
	if len(starts) != 2 || starts[0].CWD != projectRoot || starts[1].CWD != fixed {
		t.Fatalf("profile working directories = %#v", starts)
	}
}

func configuredTestProfile(t *testing.T, id string, kind ProfileKind) Profile {
	t.Helper()
	profile := Profile{
		ID:         id,
		Name:       id,
		Kind:       kind,
		Executable: filepath.Join(t.TempDir(), id),
	}
	if kind == ProfileAgent {
		profile.Provider = strings.TrimPrefix(id, "agent-")
		if profile.Provider == "" {
			profile.Provider = "test"
		}
	}
	return profile
}

File: crates/ptrack-terminal/src/profile_config_test.rs
use std::collections::BTreeMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::atomic::{AtomicU64, Ordering};

use crate::profile::{
    CwdPolicy, DEFAULT_PROFILE_THEME, ExitBehavior, MAX_PROFILE_FONT_SIZE, Profile, ProfileKind,
};
use crate::profile_config::{
    MAX_CONFIGURED_PROFILES, PROFILE_CONFIG_VERSION, ProfileConfig, load_profile_config,
    load_profile_config_if_exists, merge_profiles, profile_config_path, save_profile_config,
    validate_profile_config,
};

fn profile(id: &str, kind: ProfileKind, executable: &Path) -> Profile {
    Profile {
        id: id.to_owned(),
        name: id.to_owned(),
        kind,
        provider: if kind == ProfileKind::Agent {
            id.strip_prefix("agent-").unwrap_or("test").to_owned()
        } else {
            String::new()
        },
        executable: executable.to_string_lossy().into_owned(),
        args: Vec::new(),
        env: BTreeMap::new(),
        theme: String::new(),
        font_family: String::new(),
        font_size: 0,
        scrollback: 0,
        cwd_policy: CwdPolicy::Requested,
        fixed_cwd: String::new(),
        exit_behavior: ExitBehavior::Keep,
    }
}

#[test]
fn global_profile_config_path_is_not_project_scoped() {
    let home = Path::new("/global/ptrack-home");
    assert_eq!(
        profile_config_path(home),
        home.join("terminal-profiles.json")
    );
}

#[test]
fn config_validation_enforces_version_count_and_unique_profiles() {
    let directory = test_directory("config-validation");
    let executable = directory.join("shell");
    let valid = ProfileConfig {
        version: PROFILE_CONFIG_VERSION,
        profiles: vec![profile("shell", ProfileKind::Shell, &executable)],
    };
    let normalized = validate_profile_config(&valid).unwrap();
    assert_eq!(normalized.profiles[0].theme, DEFAULT_PROFILE_THEME);

    let mut wrong_version = valid.clone();
    wrong_version.version = 2;
    assert!(validate_profile_config(&wrong_version).is_err());
    let duplicate = ProfileConfig {
        version: PROFILE_CONFIG_VERSION,
        profiles: vec![valid.profiles[0].clone(), valid.profiles[0].clone()],
    };
    assert!(validate_profile_config(&duplicate).is_err());
    let too_many = ProfileConfig {
        version: PROFILE_CONFIG_VERSION,
        profiles: (0..=MAX_CONFIGURED_PROFILES)
            .map(|index| profile(&format!("shell-{index}"), ProfileKind::Shell, &executable))
            .collect(),
    };
    assert!(validate_profile_config(&too_many).is_err());
    remove_test_directory(&directory);
}

#[test]
fn strict_load_rejects_unknown_fields_trailing_data_empty_and_oversize() {
    let directory = test_directory("config-strict-load");
    let cases = [
        ("version", r#"{"version":2,"profiles":[]}"#.to_owned()),
        (
            "unknown-root",
            r#"{"version":1,"profiles":[],"extra":true}"#.to_owned(),
        ),
        (
            "unknown-profile",
            r#"{"version":1,"profiles":[{"id":"shell","name":"Shell","kind":"shell","executable":"/bin/sh","extra":true}]}"#.to_owned(),
        ),
        (
            "trailing",
            r#"{"version":1,"profiles":[]} {}"#.to_owned(),
        ),
        ("empty", String::new()),
        ("oversize", " ".repeat(256 * 1_024 + 1)),
    ];
    for (name, contents) in cases {
        let path = directory.join(format!("{name}.json"));
        fs::write(&path, contents).unwrap();
        assert!(
            load_profile_config(&path).is_err(),
            "accepted invalid case {name}"
        );
    }
    remove_test_directory(&directory);
}

#[test]
fn missing_config_is_the_only_nonfatal_load_result() {
    let directory = test_directory("config-missing");
    let missing = directory.join("missing.json");
    assert_eq!(load_profile_config_if_exists(&missing).unwrap(), None);

    let invalid = directory.join("invalid.json");
    fs::write(&invalid, b"{}").unwrap();
    assert!(load_profile_config_if_exists(&invalid).is_err());
    remove_test_directory(&directory);
}

#[test]
fn private_atomic_round_trip_replaces_content_without_environment_snapshot() {
    let root = test_directory("config-round-trip");
    let private_directory = root.join("new-private-home");
    let path = private_directory.join("terminal-profiles.json");
    let executable = root.join("shell");
    let mut first = profile("shell-default", ProfileKind::Shell, &executable);
    first.name = "First".to_owned();
    first.env.insert("PAGER".to_owned(), "less".to_owned());
    save_profile_config(
        &path,
        &ProfileConfig {
            version: PROFILE_CONFIG_VERSION,
            profiles: vec![first.clone()],
        },
    )
    .unwrap();

    first.name = "Second".to_owned();
    first.env.insert("PAGER".to_owned(), "more".to_owned());
    save_profile_config(
        &path,
        &ProfileConfig {
            version: PROFILE_CONFIG_VERSION,
            profiles: vec![first],
        },
    )
    .unwrap();

    let contents = fs::read_to_string(&path).unwrap();
    assert!(contents.ends_with('\n'));
    assert!(!contents.contains("PTRACK_CAPABILITY_TOKEN"));
    assert_eq!(
        fs::read_dir(&private_directory).unwrap().count(),
        1,
        "atomic save left a temporary file"
    );
    let loaded = load_profile_config(&path).unwrap();
    assert_eq!(loaded.profiles[0].name, "Second");
    assert_eq!(loaded.profiles[0].env["PAGER"], "more");
    assert_eq!(loaded.profiles[0].theme, DEFAULT_PROFILE_THEME);

    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        assert_eq!(
            fs::metadata(&path).unwrap().permissions().mode() & 0o777,
            0o600
        );
        assert_eq!(
            fs::metadata(&private_directory)
                .unwrap()
                .permissions()
                .mode()
                & 0o777,
            0o700
        );
    }
    remove_test_directory(&root);
}

#[test]
fn merge_replaces_stable_ids_adds_shells_and_preserves_agent_launch_identity() {
    let directory = test_directory("config-merge");
    let shell_executable = directory.join("shell");
    let agent_executable = directory.join("codex");
    let mut shell = profile("shell-default", ProfileKind::Shell, &shell_executable);
    shell.name = "Default shell".to_owned();
    let mut agent = profile("agent-codex", ProfileKind::Agent, &agent_executable);
    agent.args = vec!["--discovered".to_owned()];
    agent.env.insert("CODEX_MODE".to_owned(), "base".to_owned());

    let mut presentation_override = agent.clone();
    presentation_override.name = "Codex focused".to_owned();
    presentation_override.theme = "solarized-dark".to_owned();
    presentation_override.font_size = MAX_PROFILE_FONT_SIZE;
    presentation_override.exit_behavior = ExitBehavior::CloseOnSuccess;
    let custom = profile("shell-tools", ProfileKind::Shell, &shell_executable);
    let merged = merge_profiles(&[agent.clone(), shell], &[custom, presentation_override]).unwrap();
    assert_eq!(
        merged
            .iter()
            .map(|profile| profile.id.as_str())
            .collect::<Vec<_>>(),
        ["shell-default", "shell-tools", "agent-codex"]
    );
    assert_eq!(merged[2].args, agent.args);
    assert_eq!(merged[2].env, agent.env);
    assert_eq!(merged[2].theme, "solarized-dark");

    let mut changed = agent.clone();
    changed.args.push("--changed".to_owned());
    assert!(merge_profiles(&[agent], &[changed]).is_err());
    let custom_agent = profile("agent-custom", ProfileKind::Agent, &agent_executable);
    assert!(merge_profiles(&[], &[custom_agent]).is_err());
    remove_test_directory(&directory);
}

#[test]
fn merge_rejects_kind_provider_and_duplicate_repurposing() {
    let directory = test_directory("config-merge-reject");
    let executable = directory.join("codex");
    let agent = profile("agent-codex", ProfileKind::Agent, &executable);

    assert!(merge_profiles(&[], &[agent.clone(), agent.clone()]).is_err());
    let mut changed_kind = agent.clone();
    changed_kind.kind = ProfileKind::Shell;
    changed_kind.provider.clear();
    assert!(merge_profiles(std::slice::from_ref(&agent), &[changed_kind]).is_err());
    let mut changed_provider = agent.clone();
    changed_provider.provider = "other".to_owned();
    assert!(merge_profiles(std::slice::from_ref(&agent), &[changed_provider]).is_err());
    let mut changed_cwd = agent.clone();
    changed_cwd.cwd_policy = CwdPolicy::Project;
    assert!(merge_profiles(&[agent], &[changed_cwd]).is_err());
    remove_test_directory(&directory);
}

static TEST_DIRECTORY_COUNTER: AtomicU64 = AtomicU64::new(0);

fn test_directory(label: &str) -> PathBuf {
    let path = std::env::temp_dir().join(format!(
        "ptrack-terminal-{label}-{}-{}",
        std::process::id(),
        TEST_DIRECTORY_COUNTER.fetch_add(1, Ordering::Relaxed)
    ));
    fs::create_dir_all(&path).unwrap();
    path
}

fn remove_test_directory(path: &Path) {
    let _ = fs::remove_dir_all(path);
}
